vsintegration/src/FSharp.Editor/CodeFixes/ProposeUppercaseLabel.fs

```diff
@@ -3,31 +3,43 @@
 namespace Microsoft.VisualStudio.FSharp.Editor
 
 open System.Composition
-open System.Threading.Tasks
 open System.Collections.Immutable
 
 open Microsoft.CodeAnalysis.CodeFixes
-open Microsoft.CodeAnalysis.CodeActions
+open Microsoft.CodeAnalysis.Text
 
 open FSharp.Compiler.Diagnostics
 
-[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = "ProposeUpperCaseLabel"); Shared>]
-type internal ProposeUpperCaseLabelCodeFixProvider [<ImportingConstructor>] () =
-    inherit CodeFixProvider()
-
-    override _.FixableDiagnosticIds = ImmutableArray.Create("FS0053")
-
-    override _.RegisterCodeFixesAsync context : Task =
-        asyncMaybe {
-            let textChanger (originalText: string) =
-                originalText.[0].ToString().ToUpper() + originalText.Substring(1)
+open CancellableTasks
 
-            let! solutionChanger, originalText = SymbolHelpers.changeAllSymbolReferences (context.Document, context.Span, textChanger)
-
-            let title =
-                CompilerDiagnostics.GetErrorMessage(FSharpDiagnosticKind.ReplaceWithSuggestion <| textChanger originalText)
+[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.ProposeUppercaseLabel); Shared>]
+type internal ProposeUppercaseLabelCodeFixProvider [<ImportingConstructor>] () =
+    inherit CodeFixProvider()
 
-            context.RegisterCodeFix(CodeAction.Create(title, solutionChanger, title), context.Diagnostics)
-        }
-        |> Async.Ignore
-        |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+    override _.FixableDiagnosticIds = ImmutableArray.Create "FS0053"
+
+    override this.RegisterCodeFixesAsync context = context.RegisterFsharpFix this
+
+    interface IFSharpCodeFixProvider with
+        member _.GetCodeFixIfAppliesAsync context =
+            cancellableTask {
+                let! errorText = context.GetSquigglyTextAsync()
+
+                // probably not the 100% robust way to do that
+                // but actually we could also just implement the code fix for this case as well
+                if errorText.StartsWith "exception " then
+                    return None
+                else
+                    let upperCased = $"{errorText[0]}".ToUpper() + errorText.Substring(1)
```

Comment:
This is a wasteful way for effectively doing what System.Char.ToUpper(errorText[0]) does.

(seeing this, there should also be codefix for cases where interpolated string feature is used for no observed benefit. e.g. just a hole with argument, but nothing around it)

Reply:
We need the full string for the title "Replace with 'Field'". We could have it like this:
```fsharp
let upperCasedFirstLetter = $"{errorText[0]}".ToUpper()

let title =
    CompilerDiagnostics.GetErrorMessage(FSharpDiagnosticKind.ReplaceWithSuggestion (upperCasedFirstLetter + errorText.Substring(1)))

return
    (Some
        {
            Name = CodeFix.ProposeUppercaseLabel
            Message = title
            Changes = [ TextChange(TextSpan(context.Span.Start, 1), upperCasedFirstLetter) ]
        })
```
WDYT? I don't have a preference here really.

Reply:
This is the part which is doing interpolation + string operation, even when it knows for sure it just operates on a single char:
`$"{errorText[0]}".ToUpper() ` .
That should be changed, the rest is fine.

Reply:
Ah. Misread your initial comment, yep, sure.

Reply:
As for the proposed string interpol code fix - good idea, will write it down. Although it would be rather a code refactoring.